### PR TITLE
Add GAFN for hypergrid with sparse reward

### PR DESCRIPTION
**Pull request by hyeok9855**
- [x] I've read the [.github/CONTRIBUTING.md](.github/CONTRIBUTING.md) file
- [x] My code follows the typing guidelines
- [ ] I've added appropriate tests
- [x] I've run pre-commit hooks locally

## Description

Add GAFN (Generative Augmented Flow Networks, [Pan et al. 2021](https://arxiv.org/abs/2210.03308)) implementation.

This PR includes, specifically,
- GAFN implementation for HyperGrid env (see `tutorials/examples/train_hypergrid_gafn.py`)
- Make the reward function of HyperGrid configurable (see `gym/helpers/rewards.py`) and add a few reward functions. 
- Minor renaming `...dist_pmf` -> `...dist` (e.g., `true_dist_pmf` -> `true_dist`) since pmf is not general but only for discrete envs.

### CODE REVIEW

File: src/gfn/gflownet/trajectory_balance.py [hunk 1]
```diff
@@ -51,18 +52,12 @@ def __init__(
...
 
-        if isinstance(logZ, float):
-            self.logZ = nn.Parameter(torch.tensor(logZ))
-        else:
-            assert isinstance(
-                logZ, ScalarEstimator
-            ), "logZ must be either float or a ScalarEstimator"
-            self.logZ = logZ
-
+        self.logZ = logZ or nn.Parameter(torch.tensor(init_logZ))
```
[the remark this answers is not in the document]
Comment (hyeok9855) [reply src/gfn/gflownet/trajectory_balance.py:60]: This change was just a refactoring. I just wanted the parameters `logZ` to be a learnable component or None, and also want to ensure that the auto-initialization occurs only when logZ is None.

File: tutorials/examples/train_hypergrid_gafn.py [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+    Args:
+        pf: The forward policy estimator.
+    """
+
+    def __init__(
+        self,
+        pf: Estimator,
+        pb: Estimator,
+        rnd: RND,
+        logZ: nn.Parameter | ScalarEstimator | None = None,
```
[the remark this answers is not in the document]
Comment (hyeok9855) [reply tutorials/examples/train_hypergrid_gafn.py:126]: You mean the type of `logZ`? Note that I've also changed the hinting in the parent class (TBGFlowNet) accordingly.

File: tutorials/examples/train_hypergrid_gafn.py [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+            log_state_flows = torch.zeros(
+                trajectories.states.tensor.shape[:-1],
+                device=trajectories.states.device,
+            )
+            # Note: These two lines below may require a lot of memory if trajectory length
+            # is very long.
+            log_state_flows[~trajectories.states.is_sink_state] = self.flow_estimator(
+                trajectories.states[~trajectories.states.is_sink_state]
+            ).squeeze(-1)
+            edge_ri = self.rnd.compute_intrinsic_reward(trajectories.states)
```
[the remark this answers is not in the document]
Comment (hyeok9855) [reply tutorials/examples/train_hypergrid_gafn.py:197]: You're right!

File: tutorials/examples/train_hypergrid_gafn.py [hunk 4]
```diff
@@ -0,0 +1,427 @@
...
+            # Also, they define r(s_t \to s_{t+1}) = r(s_{t+1}) and here we follow
+            # this definition.
+            assert self.flow_estimator is not None
+            log_state_flows = torch.zeros(
+                trajectories.states.tensor.shape[:-1],
+                device=trajectories.states.device,
+            )
+            # Note: These two lines below may require a lot of memory if trajectory length
+            # is very long.
+            log_state_flows[~trajectories.states.is_sink_state] = self.flow_estimator(
```
[the remark this answers is not in the document]
Comment (josephdviviano) [reply tutorials/examples/train_hypergrid_gafn.py:194]: I see - so this incentivizes exploration at *every* step of the trajectory, not only among terminal states.

**Review by josephdviviano — APPROVED** [review]
Thanks for answering my questions!

File: tutorials/examples/train_hypergrid_gafn.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (hyeok9855) [reply tutorials/examples/train_hypergrid_gafn.py:198]: I made the change. But note that using `.is_sink_state` + negation `~` + indexing `__getitem__` might make this slower than before.
